Add feature extract

abdul-moize opened the pull request:
This PR adds the extract feature to master branch. Extract feature crawls through the [link](https://pakistanistores.com/prices/home-appliances/led-tv-prices) and extracts each product from all the pages. Then, one by one, it goes to the product information page and extracts description of each product. For each product it extracts `name`, `link`, `img_link`, `price`, and `description`

To test this PR follow these steps:
1. pull this branch and navigate to root directory
2. run: `pip install -r requirements.txt`
3. run: `scrapy crawl pakistan_stores -a url=https://pakistanistores.com/prices/home-appliances/led-tv-prices`

This will generate a file named `products.json`. This file will contain all the products and their relevant information.

arslanashraf7 reviewed (changes requested):
Can you make it a little more flexible e.g. How about I can give the site URL while running the command instead of baking it inside the spider?

abdul-moize commented:
> Can you make it a little more flexible e.g. How about I can give the site URL while running the command instead of baking it inside the spider?

Added the url parameter and updated the PR desciption

arslanashraf7 commented on tutorial/middlewares.py:
```diff
@@ -3,10 +3,9 @@
 # See documentation in:
```
Do we need this file?

abdul-moize replied:
Not at the moment. Removed it

arslanashraf7 commented on tutorial/settings.py:
```diff
@@ -7,82 +7,82 @@
 #     https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
```
We might not need this file as well.

abdul-moize replied:
I tried removing it but scrapy then complains that it needs this file
![image](https://user-images.githubusercontent.com/88365652/132628520-125f0f3c-4aa7-4a2f-85ec-8f1465157ddd.png)

arslanashraf7 commented on tutorial/spiders/example.py:
```diff
@@ -2,9 +2,9 @@
 
```
Same as above comment

abdul-moize replied:
Removed

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+        product_containers = response.css("li.col-md-3.col-md-3.col-sm-6.col-xs-6 a")
+        for product_container in product_containers:
+            name = product_container.attrib["title"]
+            url = response.urljoin(product_container.attrib["href"])
```
Let's name it `page_url` for better readability.

abdul-moize replied:
Done

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+        else:
+            for product in self.products:
+                link = product["link"]
+                binded_f = partial(self.extract_description, product)
```
Rename `binded_f ` it to a verbose name?

abdul-moize replied:
Done

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+            name = product_container.attrib["title"]
+            url = response.urljoin(product_container.attrib["href"])
+            img_url = product_container.css("img.lazyload").attrib["data-src"]
+            price = product_container.css("div.primary-color.price::text").get()[0:-1]
```
Can you consider two conditions here:

- What if `product_container.css("div.primary-color.price::text").get()` returns None, will it ever return None?
- Try using `[:-1]` instead of `[0:-1]` for simplicity, it will behave the same

abdul-moize replied:
Done

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+            price = product_container.css("div.primary-color.price::text").get()[0:-1]
+            product = {
+                "name": name,
+                "link": url,
```
```suggestion
                "product_link": url,
```

abdul-moize replied:
Done

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+            None
+        """
+        index = self.products.index(product)
+        description = response.css("div.light p::text")[-1].get()[1:-1]
```
How about if we want to make something reusable here? Can we define `::text` prop in some common place and everytime we use it from there?

abdul-moize replied:
I have created a file `constants.py` and defined a Constant `TEXT_PROP` and used it where i was previously using `::text`

arslanashraf7 commented on tutorial/spiders/pakistan_stores_spider.py:
```diff
@@ -0,0 +1,97 @@
+        self.products[index] = product
+        self.increment_and_check()
+
+    def write_to_file(self):
```
We have done something similar in weatherman, Can we move this to a common file probably `utils.py` and in the process, we'll make it clean and move towards single responsibility for a class?

abdul-moize replied:
Done

arslanashraf7 reviewed (changes requested):
Thanks for making the last feedback changes. Some minor comments regarding code quality. Otherwise, It's looking fine

arslanashraf7 reviewed (approved):
Looks great 👍 Thanks for making all the changes, Please squash and merge.
